src/program.rs

```diff
@@ -104,9 +102,87 @@ impl<App: Application> Program<App> {
     ) -> Result<Program<App>, Error> {
         let typed_program = untyped_program.type_check()?;
         let witness_program = typed_program.decode_witness(iter)?;
-        let finalized_program = compress_and_finalize(witness_program);
+        let finalized_program = witness_program.finalize();
         Ok(finalized_program)
     }
+
+    /// Check if the program has maximal sharing.
+    ///
+    /// This imposes the following conditions:
+    /// 1. For hidden nodes, their hash must be unique in the program.
+    /// 2. For non-hidden nodes, the triple of their IMR, source type and target type
+    ///    must be unique in the program.
+    pub fn has_maximal_sharing(&self) -> bool {
+        let mut seen_hashes = HashSet::new();
+        let mut seen_keys = HashSet::new();
+
+        for node in &self.nodes {
+            if let Term::Hidden(h) = node.term() {
+                if seen_hashes.contains(h) {
+                    return false;
+                } else {
+                    seen_hashes.insert(h);
+                }
+            } else {
+                let primary_key = (node.imr, node.source_ty().clone(), node.target_ty().clone());
+
+                if seen_keys.contains(&primary_key) {
+                    return false;
+                } else {
+                    seen_keys.insert(primary_key);
+                }
+            }
+        }
+
+        true
+    }
+
+    /// Compress the program such that it has maximal sharing (see [`Self::has_maximal_sharing`]).
+    pub fn maximize_sharing(self) -> Self {
+        let mut shared_program = Vec::new();
+        let mut shared_idx = 0;
+
+        let mut primary_key_to_shared_idx = HashMap::new();
+        let mut hash_to_shared_idx = HashMap::new();
+        let mut unshared_to_shared_idx = HashMap::new();
```

Comment:
In c010975498dee1db8d1654ea7b2fd47bf78a653b:

I really think you could avoid a lot of this messiness by not computing indices until the last minute (after sharing, after pruning).

Reply:
In fact I wonder if indices should only exist during de/serialization?

During serialization you could maintain a hashmap of `(imp, types)` to indices and then you'd also get sharing for free...this seems like a bigger refactor though and probably shouldn't be part of this PR.

Reply:
The complexity of sharing comes from the relative indices inside `Term` that need to be updated. If what a node pointed to was removed in favour of an equivalent DAG earlier on, then the node needs to point to this earlier DAG.